Restyle should reflect animations and transitions

mrobinson opened the pull request:
When doing a restyle, we should apply animations and transitions to the
new style so that it is reflected in `getComputedStyle()` and the new
style information properly cascades. This is the first part of properly
ticking animations and transitions.

This causes a couple new animations tests failures (along with many new
passes), but we currently don't have support for properly handling
animations after they have completed, so this isn't totally unexpected.

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #20116

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mrobinson committed:
Restyle should reflect animations and transitions

When doing a restyle, we should apply animations and transitions to the
new style so that it is reflected in `getComputedStyle()` and the new
style information properly cascades. This is the first part of properly
ticking animations and transitions.

This causes a couple new animations tests failures (along with many new
passes), but we currently don't have support for properly handling
animations after they have completed, so this isn't totally unexpected.

highfive commented:
Heads up! This PR modifies the following files:
 * @emilio: components/layout/animation.rs, components/style/animation.rs, components/style/matching.rs

nox commented:
r? @jdm 

I don't know anything about animations yet and I'm taking the week off starting tomorrow.

emilio commented on components/style/matching.rs:
```diff
@@ -445,7 +445,10 @@ trait PrivateMatchMethods: TElement {
         if let Some(ref mut old_values) = *old_values {
-            animation_state.compute_before_change_style::<Self>(
+            // We convert old values into `before-change-style` here.
+            let old_values = Arc::make_mut(old_values);
```
So, this will copy the style almost surely, right? It'd be nice to not do that if there are no animations running.

mrobinson replied:
Oh, nice catch! I've modified this so that the methods now take an `Arc` which will ensure that callers don't accidentally clone the style like I did in this version of the change.

emilio commented on components/style/matching.rs:
```diff
@@ -475,6 +478,20 @@ trait PrivateMatchMethods: TElement {
         }
 
+        {
+            let new_values = Arc::make_mut(new_values);
```
Same, it seems this will disable all sorts of style sharing optimizations.

mrobinson replied:
Same here. I originally went with an `is_empty()` check, but ended up doing this in the methods themselves to enable more fine-grained control over when the style is cloned.

emilio reviewed (approved):
This looks fine but it'd be nice to avoid cloning the style unconditionally. Maybe just wrapping the relevant blocks in an `if !animation_state.is_empty()` or such would work.

mrobinson commented:
Thanks for the reviews! I'm about to push a new version of this change which switches back to using `Arc<ComputedValues>` everywhere so that callers of new methods don't accidentally clone styles. It will also eliminate one more use of eager `Arc::make_mut` that existed prior to this change.

mrobinson commented:
@bors-servo r=jdm,emilio

bors-servo commented:
:pushpin: Commit c89f448 has been approved by `jdm,emilio`

<!-- @bors-servo r=jdm,emilio c89f4485218e84afea1b226ec7e949b83e60a783 -->

bors-servo commented:
:hourglass: Testing commit c89f4485218e84afea1b226ec7e949b83e60a783 with merge 1274980d131e69502fd9da268a1e35bdc4f44f5a...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/Zzs0-l_GRMS3ZFaGBfaD5A)

jdm commented:
So many new passing tests!

mrobinson commented:
@bors-servo r=jdm,emilio

Trying this again with updated expectations, but this might take a couple rounds.

bors-servo commented:
:pushpin: Commit 95adeee has been approved by `jdm,emilio`

<!-- @bors-servo r=jdm,emilio 95adeee754be128aacb5f1c3e70687a4cfffb518 -->

bors-servo commented:
:hourglass: Testing commit 95adeee754be128aacb5f1c3e70687a4cfffb518 with merge ef9bce0c7dfada8f8b4dbf8215cb00616ffa457b...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/bUrq2CvsS8W86uOdv7ryzg)

mrobinson commented:
@bors-servo r=jdm,emilio

bors-servo commented:
:pushpin: Commit 5ca2e7c has been approved by `jdm,emilio`

<!-- @bors-servo r=jdm,emilio 5ca2e7ce91088c22116082147e765dde45a8cfa1 -->

jdm commented:
It might be worth using try=wpt to shake out the failures before going through a complete merge cycle.

mrobinson commented:
@jdm Thanks for the tip. I'll give that a shot for the next round.

bors-servo commented:
:hourglass: Testing commit 5ca2e7ce91088c22116082147e765dde45a8cfa1 with merge 7a75b256c8b11e41de17addc4fbcc2be98bc0ec7...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/AR-1j8BfQ4uxNqZo_qv96w)

mrobinson committed:
Skip various CSS animation tests directories

These are all going to be flaky until we get further along with the
transition and animation improvements. The flakiness is due to the fact
that it is more likely that these tests pass now.

mrobinson commented:
@bors-servo try=wpt

Trying a new version of this branch that skips various directories of tests that are now newly flaky. Hopefully this is one of the first steps in making these tests not flaky at all, but unfortunately this change causes them to pass with much more likelihood -- increasing flakiness.

bors-servo commented:
:hourglass: Trying commit adcf89e4f7c00f2a8afa2496df52a41a14516674 with merge d0d306d484f7ed5bba14875f5efa87553032587b...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/WDW2zU14QFGLuIpaPp92Lw)
State: approved= try=True

mrobinson commented:
@bors-servo r=jdm,emilio

:crossed_fingers:

bors-servo commented:
:pushpin: Commit adcf89e has been approved by `jdm,emilio`

<!-- @bors-servo r=jdm,emilio adcf89e4f7c00f2a8afa2496df52a41a14516674 -->

bors-servo commented:
:hourglass: Testing commit adcf89e4f7c00f2a8afa2496df52a41a14516674 with merge 0d0cae68d679a44e0831823fc4b65684c5e766fc...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/EmyvpDUqQ12CjYZ79j33Qg)

mrobinson commented:
@bors-servo try=wpt

Skipped a few more directories which are flaky on Mac as well.

bors-servo commented:
:hourglass: Trying commit 44f306e90260bf735d360470c1cef7affbb0835c with merge 24c77b6db1a7a9e250ad9f858f9cf9a307575bb4...

jdm commented:
Note - try=wpt is linux-only; try=wpt-mac is mac-only.

jdm commented:
There's an ini file with a problem:
```
ParseError: Token 'group_start' doesn't equal expected type 'group_end':  line 28
  File "/repo/python/servo/testing_commands.py", line 342, in test_tidy
    tidy_failed = tidy.scan(not all_files, not no_progress, stylo=stylo, no_wpt=no_wpt)
  File "/repo/python/tidy/servo_tidy/tidy.py", line 1141, in scan
    for error in errors:
  File "/repo/python/tidy/servo_tidy/tidy.py", line 517, in check_manifest_dirs
    p = parser.parse(lines)
  File "/repo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/wptmanifest/parser.py", line 789, in parse
    return p.parse(stream)
  File "/repo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/wptmanifest/parser.py", line 540, in parse
    self.manifest()
  File "/repo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/wptmanifest/parser.py", line 564, in manifest
    self.data_block()
  File "/repo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/wptmanifest/parser.py", line 585, in data_block
    self.data_block()
  File "/repo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/wptmanifest/parser.py", line 586, in data_block
    self.eof_or_end_group()
  File "/repo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/wptmanifest/parser.py", line 591, in eof_or_end_group
    self.expect(token_types.group_end)
  File "/repo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/wptmanifest/parser.py", line 555, in expect
    "Token '{}' doesn't equal expected type '{}'".format(self.token[0], type))
```

jdm commented on tests/wpt/include.ini:
```diff
@@ -19,20 +19,34 @@ skip: true
   [CSS2]
     skip: false
+    [linebox]
+      skip:false
```
Possibly the missing space here?

mrobinson replied:
Looks like I had an extra level of indentation in the `css-align` and `css-color` sections. Should be fixed now.

mrobinson commented:
@bors-servo try=wpt-mac

It looked like the tidy precheck passed without issues, so I feel a bit safer about triggering this now.

bors-servo commented:
:hourglass: Trying commit fb9c3527be05f47ff1db6231b0f94283a15c77e9 with merge 258654efa1c8d24c81ca73ea786545cdd1f3a11a...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/ZTrwguyPSSaPWiTR0cTQ3A)

mrobinson commented:
@bors-servo r=jdm,emilio

The WPT failure is:
```
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 58: ordinal not in range(128)
```

bors-servo commented:
:pushpin: Commit fb9c352 has been approved by `jdm,emilio`

<!-- @bors-servo r=jdm,emilio fb9c3527be05f47ff1db6231b0f94283a15c77e9 -->

jdm commented:
@mrobinson For future reference, look at the filtered-wpt-errorsummary.log file instead of the live log.

mrobinson commented:
@jdm Thanks for the pointer. Looks like the real failure during the try was #25513.

bors-servo commented:
:hourglass: Testing commit fb9c3527be05f47ff1db6231b0f94283a15c77e9 with merge 5925d42c2464389ba8da169a71fc54236161d5fb...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/XnKVdCvfTL2ohsskgVyxQA)

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit fb9c3527be05f47ff1db6231b0f94283a15c77e9 with merge d1279ebdc66f8735eb29f051df365c74b0a5f1f5...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/OFQO_gEjSRmVfay7_8QZqg)
Approved by: jdm,emilio
Pushing d1279ebdc66f8735eb29f051df365c74b0a5f1f5 to master...
